README.md

```diff
@@ -0,0 +1,177 @@
+
+## xtm_queue_count
+
+Function returns current message count in queue.
+
+## xtm_queue_push_fun
+
+Function push function and its argument to the queue. This function does not
+notify the consumer thread, but only push to the queue. To notify the consumer
+thread you must call xtm_queue_notify_consumer. The less often you notify, the
+greater the performance, but the greater the latency.
+Returns 0 if queue has space. Otherwise -1 with errno set to ENOBUFS.
+
+## xtm_queue_consumer_fd
+
+Returns file descriptor, that should be watched by consumer thread to
+become readable. When it became readable, consumer should call one of
+the consumer functions: xtm_queue_pop_ptrs or xtm_queue_invoke_funs_all.
+
+## xtm_queue_producer_fd
+Return file descriptor, that should be watched by producer thread to
+become readable. When it became readable, producer may push to queue,
+being sure that there is a place in it.
+
+## xtm_queue_invoke_funs_all
+
+Function calls all functions contained in the queue.
+Return count of invoked functions on success. Otherwise -(count + 1) with errno
+set appropriately (as in write(2), since it implies write to internal fd).
+
+## xtm_queue_push_ptr
+
+Function зush pointer to the queue. This function does not notify the consumer
+thread, but only push to the queue. To notify the consumer thread you must call
+xtm_queue_notify_consumer. The less often you notify, the greater the performance,
+but the greater the latency.
+Returns 0 if queue has space. Otherwise -1 with errno set to ENOBUFS.
+
+## xtm_queue_pop_ptrs
+
+Function gets up to count elements from the queue and save them in pointer array.
+Return count of extracted messages on success. Otherwise -(count + 1) with errno
+set appropriately (as in write(2), since it implies write to internal fd).
+
+## xtm_queue_consume
+
+Function flush queue pipe, according to file descriptor passed to this function.
+Return 0 on success. Otherwise -1 with errno set appropriately (as in read(2),
+since it implies read from internal fd).
+
+Examples
+--------
+
+Full example you can see in test or perf repositories, here is only simple
+example of using library.
+
+** Allocate xtm queue with appropriate flags **
+
+```c
+unsigned flags = 0;
+flags |= XTM_QUEUE_NEED_TO_CLOSE_READFD;
+flags |= XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS;
+xtm_queue = xtm_queue_new(XTM_QUEUE_SIZE, flags);
+```
+
+** Get internal fd on consumer side and pool it in the loop **
+
+```c
+int fd = xtm_queue_consumer_fd(xtm_queue);
+fd_set readset;
+FD_ZERO(&readset);
+FD_SET(fd, &readset);
+select(fd + 1, &readset, NULL, NULL, NULL);
+```
+
+** Invoke functions in case when producer thread push functions **
+
+```c
+xtm_queue_invoke_funs_all(xtm_queue);
+```
+
+** Pop pointers to data in case when producer thread push pointers **
+
+```c
+unsigned count = xtm_queue_count(xtm_queue);
+unsigned cnt = 0;
+while (cnt < count) {
+	void *ptr_array[BATCH_COUNT_MAX];
+	unsigned popcnt = (count - cnt >= BATCH_COUNT_MAX ?
```

Comment:
Why do you need `popcnt`? Why can't always specify BATCH_COUNT_MAX in `pop_ptrs` call? Anyway it returns the number of actually popped pointers.

Reply:
I wanted to show how to limit the number of called functions exactly by the value == count.
But I thought that you are right, in reality it is not necessary, I will fix it here and in the tests

Reply:
Fixed